build.rs

```diff
@@ -130,39 +187,50 @@ fn consecutive_runs(data: &[u16]) -> (Vec<u16>, Vec<u16>) {
 fn create_ss58_registry(json: &str) -> Result<proc_macro2::TokenStream, String> {
 	let registry: Registry =
 		serde_json::from_str(json).map_err(|e| format!("json parsing error: {}", e))?;
-	registry.is_valid()?;
-
-	let mut registry = registry.registry;
 
-	let mut ordered_prefixes = registry.iter().map(|i| i.prefix).collect::<Vec<_>>();
-	ordered_prefixes.sort();
+	let (mut accounts, tokens) = registry.validate()?;
 
 	// Sort by name so that we can later binary search by network
-	registry.sort_by_key(|a| a.network.clone());
+	accounts.sort_by_key(|a| a.network.clone());
 
 	// Variables to insert into quote template:
-	let identifier: Vec<_> = registry.iter().map(|r| format_ident!("{}", r.name())).collect();
-
-	let reserved_prefixes = registry.iter().filter(|r| r.is_reserved()).map(|r| r.prefix);
+	let identifier: Vec<_> = accounts.iter().map(|r| format_ident!("{}", r.name())).collect();
 
-	let count = registry.len();
-	let prefix: Vec<_> = registry.iter().map(|r| r.prefix).collect();
+	let count = accounts.len();
+	let prefix: Vec<_> = accounts.iter().map(|r| r.prefix).collect();
 
-	let mut prefix_to_idx: Vec<_> = prefix.iter().enumerate().map(|(a, b)| (b, a)).collect();
-	prefix_to_idx.sort_by_key(|(prefix, _)| *prefix);
-	let prefix_to_idx = prefix_to_idx.iter().map(|(prefix, idx)| quote! { (#prefix, #idx) });
-
-	let name = registry.iter().map(|r| r.network.clone());
-	let desc = registry.iter().map(|r| {
+	let name = accounts.iter().map(|r| r.network.clone());
+	let desc = accounts.iter().map(|r| {
 		if let Some(website) = &r.website {
 			format!("{} - <{}>", r.display_name, website)
 		} else {
 			r.display_name.clone()
 		}
 	});
 
+	let mut prefix_to_idx: Vec<_> = prefix.iter().enumerate().map(|(a, b)| (b, a)).collect();
+	prefix_to_idx.sort_by_key(|(prefix, _)| *prefix);
+	let prefix_to_idx = prefix_to_idx.iter().map(|(prefix, idx)| quote! { (#prefix, #idx) });
+
+	let ident_to_tokens: Vec<_> = accounts
+		.iter()
+		.map(|r| {
+			let t = r.tokens(&tokens);
+			quote! { #( TokenRegistry::#t ,)* }
+		})
+		.collect();
+
+	let reserved_prefixes = accounts.iter().filter(|r| r.is_reserved()).map(|r| r.prefix);
+
+	let mut ordered_prefixes = accounts.iter().map(|i| i.prefix).collect::<Vec<_>>();
+	ordered_prefixes.sort_unstable();
 	let (prefix_starts, prefix_ends) = consecutive_runs(ordered_prefixes.as_slice());
 
+	let token_defs: Vec<_> = tokens.iter().map(|t| t.ident()).collect();
+	let token_names: Vec<_> = tokens.iter().map(|t| t.symbol.to_owned()).collect();
+	let token_docs: Vec<_> = tokens.iter().map(|t| t.doc_string()).collect();
+	let token_decimals: Vec<_> = tokens.iter().map(|t| t.decimals).collect();
```

Comment:
Do you iterate any of these multiple times? If not, you don't need to collect them and can directly use the iterator.